BCPopover/BCPopover.m

```diff
@@ -138,23 +162,42 @@ - (NSRect)popoverWindowFrame {
 - (NSRect)screenAnchorRect {
   if (self.attachedToView) {
     NSView *anchoredView = self.attachedToView.superview;
-    return [self.attachedToView.window convertRectToScreen:[anchoredView convertRect:anchoredView.bounds toView:nil]];
+    NSRect rect = [anchoredView.window convertRectToScreen:[anchoredView convertRect:anchoredView.bounds toView:nil]];
+    return [self ensureRectFitsInParentWindow:rect];
   } else
     return [self.window frame];
 }
 
 - (NSPoint)attachToPointInScreenCoordinates {
   if (self.attachedToView) {
-    NSRect screenRect = NSInsetRect([self.attachedToView convertRect:[self.attachedToView bounds] toView:nil], -6, -6);
-    NSPoint pointAtEdge = [self pointAtEdge:self.preferredEdge ofRect:screenRect];
-    NSRect converted = [self.attachedToView.window convertRectToScreen:NSMakeRect(pointAtEdge.x, pointAtEdge.y, 0, 0)];
-    return converted.origin;
+    NSRect rectInWindow = NSInsetRect([self.attachedToView convertRect:[self.attachedToView bounds] toView:nil], -6, -6);
+    NSPoint pointAtEdge = [self pointAtEdge:self.preferredEdge ofRect:rectInWindow];
+    
+    NSWindow *window = self.attachedToView.window;
+    
+    NSRect converted = [window convertRectToScreen:NSMakeRect(pointAtEdge.x, pointAtEdge.y, 0, 0)];
+    return [self ensureRectFitsInParentWindow:converted].origin;
   } else {
     NSRect windowFrame = self.window.frame;
     return NSMakePoint(NSMinX(windowFrame), NSMaxY(windowFrame));
   }
 }
 
+- (NSRect)ensureRectFitsInParentWindow:(NSRect)rect; {
```

Comment:
This is called `ensureRectFitsInParentWindow`, but that's not what it does. If it did that, it would check that the left, right, top and bottom of the rect were completely inside the window.

This only checks that the middle of the rect is somewhere in the window, and even then it only does it for the X coordinate!

That may be all we need, but if so we should maybe use a better name? 

Also the actual pinning of one rect inside another seems like the sort of thing that should be in BCFoundation along with all the other utilities.